File: scryfall-datafetcher/aws/s3.go
package aws

import (
	"bytes"
	"context"
	"encoding/json"
	"log"
	sf "scryfall-datafetcher/scryfall"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

const BUCKET = "muspelheim"
const OBJECT = "scryfall-oracle-cards.json"

func CreateS3Client() *s3.Client {
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		log.Fatalf("unable to load SDK config: %v\n", err)
	}

	return s3.NewFromConfig(cfg)
}

func UploadCardInfo(client *s3.Client, cards []sf.Card) error {
	raw, err := json.Marshal(cards)
	if err != nil {
		log.Fatalf("error marshalling card data: %v\n", err)
	}

	_, err = client.PutObject(context.TODO(), &s3.PutObjectInput{
		Bucket: aws.String(BUCKET),
		Key:    aws.String(OBJECT),
		Body:   bytes.NewReader(raw),
	})

	if err != nil {
		log.Printf("error occurred uploading to S3 bucket\n")
		return err
	}

	return nil
}

File: datafetcher/src/s3.rs
use anyhow::Result;
use aws_sdk_s3::{primitives::ByteStream, Client};

use crate::scryfall::Card;

async fn load_client() -> Client {
    let config = aws_config::load_from_env().await;
    Client::new(&config)
}

pub async fn upload_cards(cards: Vec<Card>) -> Result<()> {
    let client = load_client().await;
    let pretty = serde_json::to_string_pretty(&cards)?;
    let body = ByteStream::from(pretty.into_bytes());
    client
        .put_object()
        .bucket("muspelheim")
        .key("scryfall-oracle-cards.json")
        .body(body)
        .send()
        .await?;

    Ok(())
}
